proposals/extensions-an-evolution-of-extension-methods.md

```diff
@@ -0,0 +1,591 @@
+certainly been a consistent stream of requests for them.
+
+Over the years, the C# language design team has received many requests for *extension properties*. Initially, extension
+properties might seem straightforward but they present several challenges. The most glaring issue is that properties
+don’t have a parameter list, so there isn’t an obvious place to declare a `this` parameter. However, even if that were
+solved, C# properties cannot declare generic type parameters. Without a way to define a generic type parameter on an
+extension property, it would be impossible to declare an `IsEmpty` property for `IEnumerable<T>`. And of course, once
+those issues are addressed (along with generic property type inference and the inevitable overload resolution work),
+why would we stop at extension properties? The next step would clearly be to add instance and static generic
+properties. Given that, extension properties trigger a bit of an avalanche of design issues that lead to a much larger
+C# feature that feels a bit niche and only saves the programmer a pair of empty parentheses. This has never seemed
+worth the investment.
+
+In addition, there a long-standing request from the .NET libraries team for *static extension methods*. Extension
+methods that are accessible through member access on a type open up new API composition scenarios. Imagine how powerful
+it would be to add a package reference to a .NET project and have new static methods accessible from `string` related
+to that package's domain! Unfortunately, a natural syntax to declare a static extension method that derives from
+classic extension method syntax has proven elusive. After all, extension methods are already declared as static
+methods. Would a static extension method be `static static` or maybe require an extra attribute? Also, the `this`
+parameter wouldn’t make sense, since a static extension method wouldn’t be passed an instance of the type. Then where
+would the target type go? Clearly, a new syntax is needed to support static extension methods but what happens to
+classic extension methods? Are there two radically different syntaxes?
+
+This proposal aims to solve these issues and more through a new declaration syntax specially tailored for extension
+members.
+
+## Detailed design
+[design]: #detailed-design
+
+### The Extension Container
+
+Extension members are declared in a new type declaration called an *extension container*.
+
+```antlr
+extension_container_declaration
+    : attributes? extension_container_modifier* 'partial'? 'extensions' identifier type_parameter_list? for_clause? type_parameter_constraints_clause* extension_container_body ';'?
+    ;
+
+extension_container_modifier
+    : 'public'
+    | 'internal'
+    | unsafe_modifier   // unsafe code support
+    ;
+    
+for_clause
+    : 'for' type
+    ;
+
+extension_container_body
+    : '{' extension_member_declaration* class_member_declaration* '}'
```

Comment:
It looks like this grammar supports "0 or more extension member decls, then 0 or more class member decls"? Is it intended that all the `class_member_declaration`s appear after the `extension_member_declaration`s?

Reply:
Not intended, no. Just an artifact of being quick and dirty. Will fix, thanks!